addOns/frontendscanner/src/main/java/org/zaproxy/zap/extension/frontendscanner/FrontEndScannerProxyListener.java

```diff
@@ -0,0 +1,143 @@
+public class FrontEndScannerProxyListener implements ProxyListener {
+    private static final Logger LOGGER = Logger.getLogger(FrontEndScannerProxyListener.class);
+
+    private FrontEndScannerAPI api;
+    private FrontEndScannerOptions options;
+
+    private static final String[] CSP_HEADERS = {
+        "Content-Security-Policy", "X-Content-Security-Policy", "X-WebKit-CSP"
+    };
+
+    public FrontEndScannerProxyListener(FrontEndScannerAPI api, FrontEndScannerOptions options) {
+        this.api = api;
+        this.options = options;
+    }
+
+    @Override
+    public boolean onHttpRequestSend(HttpMessage msg) {
+        return true;
+    }
+
+    @Override
+    public boolean onHttpResponseReceive(HttpMessage msg) {
+        if (options.isEnabled() && msg.getResponseHeader().isHtml()) {
+            try {
+                String html = msg.getResponseBody().toString();
+
+                Source document = new Source(html);
+                List<Element> heads = document.getAllElements("head");
+                Element head = heads.isEmpty() ? null : heads.get(0);
+
+                if (msg.getHistoryRef() != null) {
+                    String host = msg.getRequestHeader().getHeader("host");
+                    String frontEndApiUrl =
+                            API.getInstance().getCallBackUrl(this.api, "https://" + host);
+
+                    int historyReferenceId = msg.getHistoryRef().getHistoryId();
+
+                    String injectedContent =
+                            getInjectedContent(frontEndApiUrl, head, historyReferenceId);
+
+                    OutputDocument newResponseBody = new OutputDocument(document);
+                    int insertPosition = getInsertPosition(head);
+                    newResponseBody.insert(insertPosition, injectedContent);
+
+                    msg.getResponseBody().setBody(newResponseBody.toString());
+
+                    int newLength = msg.getResponseBody().length();
+                    msg.getResponseHeader().setContentLength(newLength);
+
+                    for (String header : CSP_HEADERS) {
+                        msg.getResponseHeader().setHeader(header, null);
+                    }
+                } else {
+                    LOGGER.debug("No historyRef found in the HttpMessage.");
+                }
+            } catch (Exception e) {
+                LOGGER.error(e.getMessage(), e);
+            }
+        }
+        return true;
+    }
+
+    private int getInsertPosition(Element head) {
+        // The payload needs to be inserted in front of as many elements as possible;
+        // But still after the `<meta>` tag (if there is any).
+        if (head == null) {
+            return 0;
+        }
+
+        List<Element> headChildren = head.getChildElements();
+        int numberOfChilden = headChildren.size();
+
+        if (numberOfChilden == 0) {
+            return "<head>".length();
```

Comment:
`head.getStartTag().getEnd()`?
Could you add a test with `<head>` not at the start? e.g.:
```html
<!doctype html>
<html lang="en">

<head>
```